### PR TITLE
Sorting cleanup.

### PR DESCRIPTION
**Pull request by kylev**
Comparable/Comparator implementation now use fast Java built-ins for
primitive types rather than hand-coded versions. Also DRY some code
and simplify logic.

This maintains the "biggest stack first" notion of "ascending", but
also fixes the inconsistent/accidental reversal of sort by mod (the
mod name should sort the same direction as the item name).

Includes a fix for #3520 .

### CODE REVIEW

File: src/main/java/appeng/util/ItemSorters.java [hunk 1]
```diff
@@ -54,38 +52,30 @@ public int compare( final IAEItemStack o1, final IAEItemStack o2 )
...
-			{
-				return Platform.getItemDisplayName( o2 ).compareToIgnoreCase( Platform.getItemDisplayName( o1 ) );
-			}
-
-			return compareToIgnoreCase;
+			return applyDirection( cmp );
 		}
 	};
+
 	public static final Comparator<IAEItemStack> CONFIG_BASED_SORT_BY_SIZE = new Comparator<IAEItemStack>()
```
Comment (orod-org) [inline src/main/java/appeng/util/ItemSorters.java]: MAJOR Make this anonymous inner class a lambda [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1604)

File: src/main/java/appeng/util/ItemSorters.java [hunk 2]
```diff
@@ -172,4 +118,12 @@ public static void setDirection( final SortDir direction )
 	{
 		Direction = direction;
 	}
+
+	private static int applyDirection(int cmp) {
```
Comment (orod-org) [inline src/main/java/appeng/util/ItemSorters.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)
MINOR '(' is not followed by whitespace. [rule](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)
MINOR ')' is not preceded with whitespace. [rule](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)

File: src/main/java/appeng/util/ItemSorters.java [hunk 3]
```diff
@@ -54,38 +52,30 @@ public int compare( final IAEItemStack o1, final IAEItemStack o2 )
 		{
 			final AEItemStack op1 = (AEItemStack) o1;
 			final AEItemStack op2 = (AEItemStack) o2;
+			int cmp = op1.getModID().compareToIgnoreCase( op2.getModID() );
 
-			if( getDirection() == SortDir.ASCENDING )
+			if( cmp == 0 )
 			{
-				return this.secondarySort( op2.getModID().compareToIgnoreCase( op1.getModID() ), o1, o2 );
+				cmp = Platform.getItemDisplayName( o1 ).compareToIgnoreCase( Platform.getItemDisplayName( o2 ));
```
Comment (orod-org) [inline src/main/java/appeng/util/ItemSorters.java]: MINOR ')' is not preceded with whitespace. [rule](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)

File: src/main/java/appeng/util/ItemSorters.java [hunk 4]
```diff
@@ -39,13 +39,11 @@
...
-			{
-				return Platform.getItemDisplayName( o1 ).compareToIgnoreCase( Platform.getItemDisplayName( o2 ) );
-			}
-			return Platform.getItemDisplayName( o2 ).compareToIgnoreCase( Platform.getItemDisplayName( o1 ) );
+			final int cmp = Platform.getItemDisplayName( o1 ).compareToIgnoreCase( Platform.getItemDisplayName( o2 ) );
+			return applyDirection( cmp );
 		}
 	};
+
 	public static final Comparator<IAEItemStack> CONFIG_BASED_SORT_BY_MOD = new Comparator<IAEItemStack>()
```
Comment (orod-org) [inline src/main/java/appeng/util/ItemSorters.java]: MAJOR Make this anonymous inner class a lambda [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1604)

File: src/main/java/appeng/util/ItemSorters.java [hunk 5]
```diff
@@ -54,38 +52,30 @@ public int compare( final IAEItemStack o1, final IAEItemStack o2 )
...
-			}
-			return compareLong( o1.getStackSize(), o2.getStackSize() );
+			final int cmp = Long.compare( o2.getStackSize(), o1.getStackSize() );
+			return applyDirection( cmp );
 		}
 	};
+
 	private static IInvTweaks api;
+
 	public static final Comparator<IAEItemStack> CONFIG_BASED_SORT_BY_INV_TWEAKS = new Comparator<IAEItemStack>()
```
Comment (orod-org) [inline src/main/java/appeng/util/ItemSorters.java]: MAJOR Make this anonymous inner class a lambda [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1604)

**Comment by orod-org** [comment]
SonarQube analysis reported 30 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 16 major
* ![MINOR][MINOR] 10 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [PatternHelper.java#L234](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/helpers/PatternHelper.java#L234): The Cyclomatic Complexity of this method "isValidItemForSlot" is 11 which is greater than 10 authorized. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![CRITICAL][CRITICAL] [NetworkInventoryHandler.java#L78](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/storage/NetworkInventoryHandler.java#L78): The Cyclomatic Complexity of this method "injectItems" is 15 which is greater than 10 authorized. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [PatternHelper.java#L81](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/helpers/PatternHelper.java#L81): Move the declaration of "outTag" closer to the code that uses it. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1941)
1. ![MAJOR][MAJOR] [PatternHelper.java#L89](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/helpers/PatternHelper.java#L89): Move the declaration of "out" closer to the code that uses it. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1941)
1. ![MAJOR][MAJOR] [ToolColorApplicator.java#L310](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/items/tools/powered/ToolColorApplicator.java#L310): Remove this call to "equals"; comparisons between unrelated types always return false. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2159)
1. ![MAJOR][MAJOR] [CraftingGridCache.java#L96](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/cache/CraftingGridCache.java#L96): Make this anonymous inner class a lambda [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1604)
1. ![MAJOR][MAJOR] [CraftingGridCache.java#L107](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/cache/CraftingGridCache.java#L107): Make this anonymous inner class a lambda [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1604)
1. ![MAJOR][MAJOR] [CraftingGridCache.java#L203](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/cache/CraftingGridCache.java#L203): A "Set<ICraftingProvider>" cannot contain a "IGridHost" [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2175)
1. ![MAJOR][MAJOR] [CraftingGridCache.java#L293](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/cache/CraftingGridCache.java#L293): Extract the assignment out of this expression. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)
1. ![MAJOR][MAJOR] [CraftingGridCache.java#L487](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b8523e91977d7a2ed814af905d9591eecbb30637/src/main/java/appeng/me/cache/CraftingGridCache.java#L487): Iterate over the "entrySet" instead of the "keySet". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2864)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by yueh** [comment]
Out of curiosity, are there any actual performance improvements as the java version is pretty much identical but simply did not exist in Java6. And still not the fastest way.

Descending by stacksize is simply the more common approach, forcing the player to constantly switch it also is a bit overkill.

Not sure about the mod id, it might be inconsistent to minecraft/vanilla to the top at some point in the past.

The goal is to refactor the whole GUI/container code, so I did not bother with it in the other PR. It is actually more like 2 PRs in one. The first one decoupling non gui code from actual GUI code and the other one making the GUI part a bit nicer. But I will most likely postpone it to rv6.

**Comment by kylev** [comment]
I wouldn't wager any performance improvements here; slow-downs aren't going to be in the int-comparison semantics. This is just janitorial work intended to do things like reduce complexity, use built-ins where possible, and clean-up unused or un-necessary bits. The intent is to make it look like more readable, every-day Java when it can. That will hopefully be less cognitive load on maintainers (who can focus on features, not clean-up) and an increasingly inviting codebase for future OSS bug fixers and contributors.

Yeah, I realized the meaning was "inverted" when I looked at the interface, but it's seems like the right choice given that people seem to want to see the things they have the *most* of. But maybe I watch too many Skyblocks. I can add a change to include Javadocs to call out exactly this reasoning. It's the sort of thing that would help future code-readers to understand that it isn't a typo or un-reported bug.

I thought that about the mod_id too, but it's the string "ID". I added the Yoyo mod just to make sure. It's not "minecraft first", it's reversed. It only looks "minecraft first" in the dev environment without other mods in play, making the bug easy to overlook. Current codebase goes from alpha-by-name to reverse-alpha-by-mod when you click on "Sort order"; this patch makes both of them alphabetical.

**Comment by kylev** [comment]
I can actually open a bug for the reversed sorting-by-mod GUI thing. It's actually a UI interaction concept really close to the one that you pointed out about with the "biggest first" thing: it's what most people expect.

Currently it does the unexpected switch from ascending to descending when going from item name to mod name.

**Comment by yueh** [comment]
Yeah, the changes are nice from a maintenance point of view. The performance part is just curiosityas I know of a couple hotspots in AE2, where it actually made some impressive improvements by optimizing it a bit with early returns or similar changes. But it was mostly because the method gets called a couple 100 million times in a short amount of time (less than a minute of sampling data).

I have absolutely no idea about the mod sorting. Like never ever used it before and it is still an 1.7 or maybe even 1.6 artifact. Maybe the mod id was different then. A bug report would certainly be nice, just to not forget it later.

A short notice before is probably the best case in most cases. Simply avoids that anyone waste their time with fixing something and then I have to reject it because it does not take some implications into account or opens new bugs, while explaining them pretty much takes as much time as fixing it myself.

**Comment by kylev** [comment]
Filed as #3520 with quick repro steps.

**Comment by orod-org** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**